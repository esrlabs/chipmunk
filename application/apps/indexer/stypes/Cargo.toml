[package]
name = "stypes"
description = "Shared types"
version = "0.1.0"
edition = "2024"

[features]
test_and_gen = []
rustcore = [
    "dep:tokio",
    "dep:walkdir",
    "dep:regex",
    "dep:envvars",
    "dlt-core/fibex",
    "dep:anyhow",
    "dep:tokio-util",
    "dlt-core/statistics",
    "dlt-core/serialization",
]
nodejs = [
    "dep:node-bindgen"
]

[dependencies]
serde = { workspace = true , features = ["derive"] }
dlt-core = { workspace = true, features = ["fibex", "serialization"] }
regex = { workspace = true, optional = true }
bincode = "1.3"
extend = { path = "../tools/extend"}
uuid = { workspace = true, features = ["serde"] }
tokio = { workspace = true, optional = true }
node-bindgen = { git = "https://github.com/infinyon/node-bindgen.git", branch="master", optional = true}
thiserror.workspace = true
walkdir = { workspace = true, optional = true }
envvars = { workspace = true, optional = true }
anyhow = { workspace = true, optional = true }
tokio-util = { workspace = true , features = ["full"], optional = true }

[dev-dependencies]
tokio = { workspace = true  }
walkdir = { workspace = true }
node-bindgen = { git = "https://github.com/infinyon/node-bindgen.git", branch="master"  }
proptest = { workspace = true }
paste = "1.0"
uuid = { workspace = true, features = ["serde", "v4"] }
remove_dir_all = "1.0"
ts-rs = { version = "10.1", features = ["uuid-impl"] }

