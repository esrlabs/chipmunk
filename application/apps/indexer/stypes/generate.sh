#!/bin/bash

echo "
This script generates TypeScript types from Rust types by executing special unit tests.
After running these tests, the types will be generated, and existing types will be overwritten
in the 'bindings' directory.
Please copy only the newly generated types to the 'application/platform/types' directory.

Notes:
* The generated files in 'bindings' are not part of the communication protocol; only the types in
  'platform' are part of the protocol.
* Please commit only the new changes and exclude the overwritten versions of existing types.
"
read -p "Do you want to continue? (y/N): " response

response=${response,,}
if [[ "$response" != "y" ]]; then
  echo "Operation aborted."
  exit 1
fi

echo "Generating types"
cargo test --release --features "test_and_gen" -- --nocapture
