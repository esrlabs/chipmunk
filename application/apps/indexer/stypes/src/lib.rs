/// The `stypes` crate provides data types used at the `rustcore` level and passed to clients.
/// While the `stypes` crate does not impose restrictions on the client type, some features are
/// specifically defined to support certain types of clients. This does not mean that `stypes` cannot
/// be used with other client types.
///
/// ## Features
/// - `nodejs`: Includes the implementation of the `TryIntoJs` trait, required for transferring data
///   into the Node.js context when using the `node_bindgen` crate.
/// - `rustcore`: Includes additional utilities and extensions for using `stypes` within the
///   `indexer` crate group.
///
/// ## Proptest Integration
/// The crate includes tests based on `proptest`. These tests not only validate the crate itself
/// but also generate binary files containing variations of each message in a predefined directory.
/// These files can later be used to test encoding/decoding on the client side.
///
/// The `test_msg` macro is used to generate tests for specific data types. For example:
///
/// ```ignore
/// test_msg!(ObserveOptions, 100);
/// ```
///
/// The above code generates 100 variations of `ObserveOptions`. During tests, the generated data
/// is saved to files in the specified path. To set the output path, use the `CHIPMUNK_PROTOCOL_TEST_OUTPUT`
/// environment variable:
///
/// ```ignore
/// export CHIPMUNK_PROTOCOL_TEST_OUTPUT="/tmp/test_data"
/// cargo test --release -- --nocapture
/// ```
///
/// If `CHIPMUNK_PROTOCOL_TEST_OUTPUT` is not set, the default path `$TMP/stypes_test` will be used.
/// It is recommended to run tests with the `--release` flag to speed up random variation generation,
/// as the process is significantly slower in debug mode.
///
/// Each data type will have its own directory, and each variation will be stored in files with
/// sequential names (`1.bin`, `2.bin`, etc.).
///
/// ## WARNING
/// When tests are run, the folder specified in `CHIPMUNK_PROTOCOL_TEST_OUTPUT` is completely deleted.
/// Be extremely cautious when setting the value of this environment variable.
///
/// ## Limitations
/// The current version of `stypes` uses `bincode` for encoding and decoding types. `bincode` requires
/// both serialization and deserialization implementations. However, using custom `serde` attributes
/// may lead to protocol instability, especially during decoding. For instance, the attribute
/// `#[serde(tag = "type", content = "value")]` makes decoding messages with these settings impossible.
/// Unfortunately, `bincode` does not raise compile-time or serialization-time errors, but only fails
/// during decoding. Therefore, it is strongly recommended to test encoding/decoding when using additional
/// attributes on types.
///
/// ## Implementation of `encode` and `decode` Methods
/// The `encode` and `decode` methods are added to each declared data type using the `encode_decode`
/// macro from the `extend` crate. This macro implements encoding and decoding for the data type
/// using `bincode`.
///
/// For example, the following code:
///
/// ```ignore
/// #[derive(Debug, Serialize, Deserialize, Clone)]
/// #[extend::encode_decode]
/// pub struct Notification {
///     pub severity: Severity,
///     pub content: String,
///     pub line: Option<usize>,
/// }
/// ```
///
/// Is transformed into:
///
/// ```ignore
/// #[derive(Debug, Serialize, Deserialize, Clone)]
/// pub struct Notification {
///     pub severity: Severity,
///     pub content: String,
///     pub line: Option<usize>,
/// }
///
/// impl Notification {
///     pub fn encode(&self) -> Result<Vec<u8>, String> {
///         bincode::serialize(self).map_err(|e| e.to_string())
///     }
///     
///     pub fn decode(buf: &[u8]) -> Result<Self, String> {
///         bincode::deserialize(buf).map_err(|e| e.to_string())
///     }
/// }
/// ```
#[cfg(test)]
mod tests;

#[cfg(feature = "nodejs")]
mod nodejs;

mod attachment;
mod callback;
mod command;
mod error;
mod lf_transition;
mod miscellaneous;
mod observe;
mod operations;
mod options;
mod plugins;
mod progress;

pub use attachment::*;
pub use callback::*;
pub use command::*;
pub use error::*;
pub use lf_transition::*;
pub use miscellaneous::*;
pub use observe::*;
pub use operations::*;
pub use options::*;
pub use plugins::*;
pub use progress::*;

pub(crate) use serde::{de::DeserializeOwned, Deserialize, Serialize};
pub(crate) use std::{collections::HashMap, path::PathBuf};
#[cfg(all(test, feature = "test_and_gen"))]
pub(crate) use ts_rs::TS;
pub(crate) use uuid::Uuid;

#[cfg(feature = "nodejs")]
pub(crate) use node_bindgen::{
    core::{safebuffer::SafeArrayBuffer, val::JsEnv, NjError, TryIntoJs},
    sys::napi_value,
};

#[cfg(test)]
pub(crate) use proptest::prelude::*;
#[cfg(all(test, feature = "test_and_gen"))]
pub(crate) use tests::*;

#[cfg(feature = "rustcore")]
pub fn serialize<T: Serialize>(v: &T) -> Result<Vec<u8>, bincode::Error> {
    bincode::serialize(v)
}
