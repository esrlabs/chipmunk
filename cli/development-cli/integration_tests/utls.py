"""
Utilities function to be used among the modules to test Chipmunk Build CLI Tool
"""

from pathlib import Path
import subprocess


def get_root() -> Path:
    """Get and validate the root directory of chipmunk repository

    Raises:
        SystemError: If validation of root directory fails

    Returns:
        Root directory of chipmunk
    """

    # We are using the utls file with the assumption that is path is `root/cli/development-cli/integration_tests`
    root_dir = Path(__file__).parent.parent.parent.parent

    # Root Dir checks. This checks depends on the current chipmunk directories' structure
    sub_dirs = (
        root_dir.joinpath(dir_name)
        for dir_name in ["application", "developing", "cli", "scripts"]
    )

    if any(not path.exists() for path in sub_dirs):
        raise SystemError(f"Root directory verification fail. Root Dir: {root_dir}")

    return root_dir


def run_command(command_args: list[str]):
    """Runs the commands and its arguments after printing it to stdout

    Args:
        command_args: The command and its arguments in one string list
    """
    command_txt = " ".join(command_args)
    print_bold(f"Command: {command_txt}")
    subprocess.run(command_args, check=True)


class bcolors:
    """Color codes representation in ANSI"""

    HEADER = "\033[95m"
    OKBLUE = "\033[94m"
    OKCYAN = "\033[96m"
    OKGREEN = "\033[92m"
    WARNING = "\033[93m"
    FAIL = "\033[91m"
    ENDC = "\033[0m"
    BOLD = "\033[1m"
    UNDERLINE = "\033[4m"


def print_bold(text: str):
    """Prints the given text to stdout with bold attribute"""
    print(f"{bcolors.BOLD}{text}{bcolors.ENDC}")


def print_blue_bold(text: str):
    """Prints the given text to stdout with bold and blue attribute"""
    print(f"{bcolors.BOLD}{bcolors.OKBLUE}{text}{bcolors.ENDC}{bcolors.ENDC}")


def print_green_bold(text: str):
    """Prints the given text to stdout with bold and green attribute"""
    print(f"{bcolors.BOLD}{bcolors.OKGREEN}{text}{bcolors.ENDC}{bcolors.ENDC}")


def print_red_bold(text: str):
    """Prints the given text to stdout with bold and red attribute"""
    print(f"{bcolors.BOLD}{bcolors.FAIL}{text}{bcolors.ENDC}{bcolors.ENDC}")


def print_cyan(text: str):
    """Prints the given text to stdout with cyan color attribute"""
    print(f"{bcolors.OKCYAN}{text}{bcolors.ENDC}")
