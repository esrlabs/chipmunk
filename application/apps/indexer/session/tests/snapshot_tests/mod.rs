//! Snapshot testing is set here using the crate `insta` and optionally its CLI tool `cargo-insta`.
//!
//! Each test will created a snapshot file for its results, which will be saved with the path
//! './snapshots/{test_name}.snap'. These files will be used as reference for future tests then
//! tests will fail on changes.
//! Changes can be seen in our build CLI tool and can be accepted via CLI arguments, however it's
//! better to use the CLI tool `cargo insta` because it's output is colored and more clear compared
//! to the build CLI tool and it provides the possibility to review and accept each change one by
//! one, on the contrary to our build tool than can only accept all changes at once.
//!
//! Snapshot files are parts of the repository and valid changes in them must be committed as well.
//!
//! Running the tests via `cargo test` will create temporary files with the name pattern
//! '{name}.snap.new'. These files are used by `cargo-insta` tool and they aren't part of the
//! repo history.
//! Running the tests via `cargo chipmunk test` will avoid creating these temporary files.

mod utls;

use std::path::PathBuf;
use stypes::{SessionAction, SessionSetup};
use utls::*;

#[tokio::test]
async fn observe_dlt_session() {
    let setup = SessionSetup {
        origin: SessionAction::File(PathBuf::from(
            "../../../developing/resources/attachments.dlt",
        )),
        parser: parsers::dlt::descriptor::get_default_options(None, None),
        source: sources::binary::raw::get_default_options(),
    };
    let session_main_file = run_observe_session(setup.clone()).await;

    let session_files = SessionFiles::from_session_file(&session_main_file);

    insta::with_settings!({
        description => "Snapshot for DLT file with text attachments.",
        info => &setup,
        omit_expression => true,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_yaml_snapshot!(session_files);
    });
}

#[tokio::test]
async fn observe_dlt_with_someip_session() {
    let setup = SessionSetup {
        origin: SessionAction::File(PathBuf::from("../../../developing/resources/someip.dlt")),
        parser: parsers::dlt::descriptor::get_default_options(
            Some(vec![PathBuf::from(
                "../../../developing/resources/someip.xml",
            )]),
            None,
        ),
        source: sources::binary::raw::get_default_options(),
    };
    let session_main_file = run_observe_session(setup.clone()).await;

    let session_files = SessionFiles::from_session_file(&session_main_file);

    insta::with_settings!({
        description => "Snapshot for DLT file with SomeIP network trace.",
        info => &setup,
        omit_expression => true,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_yaml_snapshot!(session_files);
    });
}

#[tokio::test]
async fn observe_someip_bcapng_session() {
    let setup = SessionSetup {
        origin: SessionAction::File(PathBuf::from("../../../developing/resources/someip.pcapng")),
        parser: parsers::someip::descriptor::get_default_options(Some(vec![PathBuf::from(
            "../../../developing/resources/someip.xml",
        )])),
        source: sources::binary::pcap::ng::get_default_options(),
    };

    let session_main_file = run_observe_session(setup.clone()).await;

    let session_files = SessionFiles::from_session_file(&session_main_file);

    insta::with_settings!({
        description => "Snapshot for SomeIP file with Pcapng byte source.",
        info => &setup,
        omit_expression => true,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_yaml_snapshot!(session_files);
    });
}

#[tokio::test]
async fn observe_someip_legacy_session() {
    let setup = SessionSetup {
        origin: SessionAction::File(PathBuf::from("../../../developing/resources/someip.pcap")),
        parser: parsers::someip::descriptor::get_default_options(Some(vec![PathBuf::from(
            "../../../developing/resources/someip.xml",
        )])),
        source: sources::binary::pcap::legacy::get_default_options(),
    };

    let session_main_file = run_observe_session(setup.clone()).await;

    let session_files = SessionFiles::from_session_file(&session_main_file);

    insta::with_settings!({
        description => "Snapshot for SomeIP file with Pcap Legacy byte source.",
        info => &setup,
        omit_expression => true,
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_yaml_snapshot!(session_files);
    });
}
