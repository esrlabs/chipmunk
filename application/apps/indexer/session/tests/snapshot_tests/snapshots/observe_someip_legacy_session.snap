---
source: session/tests/snapshot_tests/mod.rs
assertion_line: 114
description: Snapshot for SomeIP file with Pcap Legacy byte source.
info:
  origin:
    File: "../../../developing/resources/someip.pcap"
  parser:
    fields:
      - id: SOMEIP_PARSER_FIELD_FIBEX_FILES
        value:
          Files:
            - "../../../developing/resources/someip.xml"
    uuid: 02020202-0202-0202-0202-020202020202
  source:
    fields: []
    uuid: 10101010-1010-1010-1010-101010101010
---
session_file:
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656786973,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656786973,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656787976,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656787976,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656788976,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656788977,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656789980,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656789980,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656790983,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656790983,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656791988,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656791988,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656792992,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656792992,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656793993,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656793993,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656794996,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656794996,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656795996,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656795996,\u0006}"
  - "SD\u000465535\u000433024\u000460\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Offer 123 v1.0 Inst 1 Ttl 3 UDP 192.168.178.58:30000 TCP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00040\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30000"
  - "SD\u000465535\u000433024\u000448\u00041\u00040\u00041\u00042\u00040\u0004Flags [C0], Subscribe 123-321 v1 Inst 1 Ttl 3 UDP 192.168.178.58:30001"
  - "RPC\u0004123\u000432773\u000416\u00040\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656796996,\u0006}"
  - "RPC\u0004123\u000432773\u000416\u00041\u00040\u00041\u00042\u00040\u0004TestService::timeEvent {\u0006\ttimestamp (INT64) : 1683656796996,\u0006}"
attachments: []
